### ISSUE TITLE
How to handle muliple hubots on the same ST2 instance

### ISSUE THREAD

**Issue by jjm** [StackStorm/st2#2515]
A question following AHH#11.

How should ST2 handle multiple instances of hubot? Our ST2 installation has two hubots attached (both using the same chatops user), our own hubot that pre-dates us installing ST2 and the docker container version. If we run a `chatops.post_message` to a user, they receive an message from both bots as a DM (in slack) and I would expect if both bots were in the same channel the messages would be duplicated. However if your talking to a single Bot as a DM the messages are routed correctly.

The question is how to handle this? I think one option is to only have a single bot, but it would be nice to attache aliases to different bots (e.g. lab, dev & prod) that act on different environments via the same ST2 instance.

**Comment by emedvedev** [comment]
@jjm: yes, right now you can only have a single bot per stackstorm instance. `post_message` sends a `chatops` announcement which all bots are supposed to act upon. So the only legit solution that doesn't involve hacking is having multiple stackstorm instances.

We'll keep the issue open, because we can probably append hubot's user id to the announcement later, or do something similar to it, even if they still will only support the same set of commands.

**Comment by jjm** [comment]
Cheers, I'll deal with our issue by turning off the second bot and only turning it on for capturing blog images.

**Comment by murgurgula** [comment]
The multiclient feature would be usefull not only for hubot, but for stuff like ansible, puppet, chef and other.
In some cases due to politics/security/etc, a separation of environments "per customer" is done, so the company ends up having multiple saltstack servers (1 per client), which they want to plug to single stackstorm.

**Comment by roberterdin** [comment]
+1
this would make multi tenant chatops a lot easier!

**Comment by mf2012** [comment]
Does anyone know if this feature as been implemented? It's mid 2019 and no news since 2016

**Comment by LizhangX** [comment]
This feature would be a very much appreciated as we might want to support slack and ms teams together. Or did anyone have success with it?

**Comment by moet78** [comment]
Near the end of 2020, checking again.. Is this possible now?  we are also looking at supporting both slack and ms team.

**Comment by nzlosh** [comment]
This is possible with the err-stackstorm project.  https://err-stackstorm.readthedocs.io/en/latest/configuration.html#route-key

**Comment by jjm** [comment]
@StackStorm/maintainers @StackStorm/contributors Should we close this one off? I raised it around 5 years ago and there's ongoing discussions on the future of hubot.

_Just reviewing and cleaning up my still open raised issues._

**Comment by cognifloyd** [comment]
TL;DR; Yes. I think we can close this.

As the implementation for multi-bot routing will depend on which bot backend is used, I think we can close this _for hubot_ as hubot itself is on the way out.

Err-stackstorm already has this feature (thanks @nzlosh!). afaik, that uses the routing keys that ST2 core already makes available. Unlike Errbot, our Hubot integration only used the routing key to distinguish between chat providers (slack, irc, etc), not individual bots.

Once we replace Hubot with OpsDroid (no integration exists yet), OpsDroid also supports naming each chat service connection, so it should be able to use the route key flexibly like Errbot does.
